### PR TITLE
Gracefully handle connection close in SocketOutput.ProducingComplete

### PR DESCRIPTION
**Pull request by halter73**

### CODE REVIEW

File: test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs [hunk 1]
```diff
@@ -544,5 +537,37 @@ public void ProducingStartAndProducingCompleteCanBeUsedDirectly()
...
+                    default(ArraySegment<byte>), default(CancellationToken), socketDisconnect: true);
+
+                Assert.True(connection.SocketClosed.Wait(1000));
+
+                var start = socketOutput.ProducingStart();
+
+                Assert.True(start.IsDefault);
+                // ProducingComplete should not throw given a default iterator
+                socketOutput.ProducingComplete(start);
+
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs]: nit: remove blank line

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketOutput.cs [hunk 2]
```diff
@@ -241,7 +241,10 @@ public MemoryPoolIterator ProducingStart()
 
         public void ProducingComplete(MemoryPoolIterator end)
         {
-            Debug.Assert(!_lastStart.IsDefault);
+            if(_lastStart.IsDefault)
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketOutput.cs]: nit: space after `if`

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketOutput.cs [hunk 3]
```diff
@@ -535,7 +539,6 @@ private class WriteContext
 
             public int WriteStatus;
             public Exception WriteError;
-            public int ShutdownSendStatus;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketOutput.cs]: Why do we have these as fields instead of properties?

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketOutput.cs]: I don't really know. I think the perf would be the same. Brevity I guess. At least it's confined to a private class. I'm not going to change it for this PR.

**Comment by cesarbs** [comment]
:ship:

**Pull request merged**